File: heap-vs-stack/comprehensive_usage.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <iomanip>

// Helper to print memory addresses
void printAddr(const char* label, void* addr) {
    std::cout << "  " << std::setw(40) << label << ": 0x" 
              << std::hex << reinterpret_cast<uintptr_t>(addr) << std::dec << std::endl;
}

class DataObject {
    int id;
    char buffer[100];
public:
    DataObject(int i) : id(i) {
        std::cout << "  Constructor called for object " << id << std::endl;
    }
    ~DataObject() {
        std::cout << "  Destructor called for object " << id << std::endl;
    }
    int getId() const { return id; }
};

// Example 1: Stack allocation - use and cleanup
void stackAllocationExample() {
    std::cout << "\n=== 1. Stack Allocation - Allocation, Use, Deallocation ===" << std::endl;
    
    std::cout << "Before allocation:" << std::endl;
    
    int x = 42;
    std::cout << "After allocation:" << std::endl;
    printAddr("Variable x (stack)", &x);
    
    std::cout << "Using variable:" << std::endl;
    std::cout << "  x = " << x << std::endl;
    x = 100;
    std::cout << "  x after modification = " << x << std::endl;
    
    DataObject obj(1);
    printAddr("Object on stack", &obj);
    std::cout << "Using object: id = " << obj.getId() << std::endl;
    
    std::cout << "End of function - automatic cleanup..." << std::endl;
}

// Example 2: Heap allocation - manual memory management
void heapAllocationExample() {
    std::cout << "\n=== 2. Heap Allocation - Manual Memory Management ===" << std::endl;
    
    std::cout << "Before allocation:" << std::endl;
    
    int* ptr = new int(42);
    std::cout << "After allocation:" << std::endl;
    printAddr("Pointer 'ptr' itself (stack)", &ptr);
    printAddr("Data pointed to (heap)", ptr);
    
    std::cout << "Using heap variable:" << std::endl;
    std::cout << "  *ptr = " << *ptr << std::endl;
    *ptr = 100;
    std::cout << "  *ptr after modification = " << *ptr << std::endl;
    
    DataObject* objPtr = new DataObject(2);
    printAddr("Object pointer (stack)", &objPtr);
    printAddr("Object on heap", objPtr);
    std::cout << "Using heap object: id = " << objPtr->getId() << std::endl;
    
    std::cout << "Manual cleanup:" << std::endl;
    delete ptr;
    std::cout << "  Deleted int" << std::endl;
    delete objPtr;
    
    std::cout << "End of function - stack cleaned, heap manually freed" << std::endl;
}

// Example 3: Smart pointers - automatic heap cleanup
void smartPointerExample() {
    std::cout << "\n=== 3. Smart Pointers - Automatic Heap Management ===" << std::endl;
    
    std::cout << "Creating unique_ptr:" << std::endl;
    auto ptr = std::make_unique<int>(42);
    printAddr("unique_ptr (stack)", &ptr);
    printAddr("Data (heap)", ptr.get());
    
    std::cout << "Using unique_ptr:" << std::endl;
    std::cout << "  *ptr = " << *ptr << std::endl;
    *ptr = 100;
    std::cout << "  *ptr after modification = " << *ptr << std::endl;
    
    std::cout << "Creating unique_ptr to object:" << std::endl;
    auto objPtr = std::make_unique<DataObject>(3);
    std::cout << "Using object: id = " << objPtr->getId() << std::endl;
    
    std::cout << "End of function - automatic cleanup..." << std::endl;
}

// Example 4: Function call stack demonstration
void functionC() {
    std::cout << "  In functionC" << std::endl;
    int c = 3;
    printAddr("Variable c in functionC", &c);
    std::cout << "  functionC returning..." << std::endl;
}

void functionB() {
    std::cout << "  In functionB" << std::endl;
    int b = 2;
    printAddr("Variable b in functionB", &b);
    functionC();
    std::cout << "  Back in functionB, b still valid: " << b << std::endl;
    std::cout << "  functionB returning..." << std::endl;
}

void functionA() {
    std::cout << "  In functionA" << std::endl;
    int a = 1;
    printAddr("Variable a in functionA", &a);
    functionB();
    std::cout << "  Back in functionA, a still valid: " << a << std::endl;
    std::cout << "  functionA returning..." << std::endl;
}

void functionCallExample() {
    std::cout << "\n=== 4. Function Call Stack ===" << std::endl;
    std::cout << "Calling nested functions (watch stack addresses):" << std::endl;
    functionA();
    std::cout << "All functions returned, stack unwound" << std::endl;
}

// Example 5: Array allocation comparison
void arrayAllocationExample() {
    std::cout << "\n=== 5. Array Allocation Comparison ===" << std::endl;
    
    std::cout << "Stack array:" << std::endl;
    int stackArray[5] = {1, 2, 3, 4, 5};
    printAddr("stackArray (stack)", stackArray);
    std::cout << "  Using: stackArray[0] = " << stackArray[0] << std::endl;
    stackArray[0] = 99;
    std::cout << "  After modification: stackArray[0] = " << stackArray[0] << std::endl;
    
    std::cout << "\nHeap array:" << std::endl;
    int* heapArray = new int[5]{1, 2, 3, 4, 5};
    printAddr("heapArray pointer (stack)", &heapArray);
    printAddr("heapArray data (heap)", heapArray);
    std::cout << "  Using: heapArray[0] = " << heapArray[0] << std::endl;
    heapArray[0] = 99;
    std::cout << "  After modification: heapArray[0] = " << heapArray[0] << std::endl;
    
    std::cout << "\nVector (heap-managed):" << std::endl;
    std::vector<int> vec = {1, 2, 3, 4, 5};
    printAddr("vector object (stack)", &vec);
    printAddr("vector data (heap)", vec.data());
    std::cout << "  Using: vec[0] = " << vec[0] << std::endl;
    vec[0] = 99;
    std::cout << "  After modification: vec[0] = " << vec[0] << std::endl;
    
    std::cout << "\nCleanup:" << std::endl;
    delete[] heapArray;
    std::cout << "  Heap array deleted manually" << std::endl;
    std::cout << "  Stack array and vector auto-cleaned at scope end" << std::endl;
}

// Example 6: Memory reuse demonstration
void memoryReuseExample() {
    std::cout << "\n=== 6. Memory Reuse Patterns ===" << std::endl;
    
    void* addr1;
    void* addr2;
    
    {
        std::cout << "First scope:" << std::endl;
        int x = 10;
        addr1 = &x;
        printAddr("Variable x in first scope", &x);
        std::cout << "  x = " << x << std::endl;
    }
    
    {
        std::cout << "\nSecond scope (stack reused):" << std::endl;
        int y = 20;
        addr2 = &y;
        printAddr("Variable y in second scope", &y);
        std::cout << "  y = " << y << std::endl;
        
        if (addr1 == addr2) {
            std::cout << "  Notice: Same stack location reused!" << std::endl;
        }
    }
    
    std::cout << "\nHeap doesn't reuse like stack:" << std::endl;
    int* p1 = new int(10);
    printAddr("First heap allocation", p1);
    delete p1;
    
    int* p2 = new int(20);
    printAddr("Second heap allocation", p2);
    std::cout << "  Heap addresses may differ" << std::endl;
    delete p2;
}

int main() {
    std::cout << "=== Comprehensive Memory Usage Patterns ===" << std::endl;
    std::cout << "Demonstrating allocation -> use -> deallocation for each pattern\n" << std::endl;
    
    stackAllocationExample();
    heapAllocationExample();
    smartPointerExample();
    functionCallExample();
    arrayAllocationExample();
    memoryReuseExample();
    
    std::cout << "\n=== Summary ===" << std::endl;
    std::cout << "Stack: Allocate -> Use -> Auto-cleanup (LIFO)" << std::endl;
    std::cout << "Heap:  Allocate -> Use -> Manual cleanup (or smart pointers)" << std::endl;
    std::cout << "       Heap persists across function calls until freed" << std::endl;
    
    return 0;
}

File: heap-vs-stack/comprehensive_usage.rs
use std::rc::Rc;

fn print_addr<T>(label: &str, ptr: &T) {
    println!("  {:40}: {:p}", label, ptr);
}

struct DataObject {
    id: i32,
}

impl DataObject {
    fn new(id: i32) -> Self {
        println!("  Constructor called for object {}", id);
        DataObject { id }
    }
    
    fn get_id(&self) -> i32 {
        self.id
    }
}

impl Drop for DataObject {
    fn drop(&mut self) {
        println!("  Destructor called for object {}", self.id);
    }
}

// Example 1: Stack allocation - use and cleanup
fn stack_allocation_example() {
    println!("\n=== 1. Stack Allocation - Allocation, Use, Deallocation ===");
    
    println!("Before allocation:");
    
    let x = 42;
    println!("After allocation:");
    print_addr("Variable x (stack)", &x);
    
    println!("Using variable:");
    println!("  x = {}", x);
    let x = 100;  // Shadowing
    println!("  x after shadowing = {}", x);
    
    let obj = DataObject::new(1);
    print_addr("Object on stack", &obj);
    println!("Using object: id = {}", obj.get_id());
    
    println!("End of function - automatic cleanup...");
}

// Example 2: Heap allocation - Box
fn heap_allocation_example() {
    println!("\n=== 2. Heap Allocation - Box (Automatic Management) ===");
    
    println!("Before allocation:");
    
    let ptr = Box::new(42);
    println!("After allocation:");
    print_addr("Box 'ptr' itself (stack)", &ptr);
    print_addr("Data pointed to (heap)", &*ptr);
    
    println!("Using heap variable:");
    println!("  *ptr = {}", *ptr);
    let mut ptr = ptr;
    *ptr = 100;
    println!("  *ptr after modification = {}", *ptr);
    
    let obj_ptr = Box::new(DataObject::new(2));
    print_addr("Box pointer (stack)", &obj_ptr);
    print_addr("Object on heap", &*obj_ptr);
    println!("Using heap object: id = {}", obj_ptr.get_id());
    
    println!("End of function - automatic cleanup...");
}

// Example 3: Ownership transfer
fn ownership_transfer_example() {
    println!("\n=== 3. Ownership Transfer ===");
    
    println!("Creating owned value:");
    let s1 = String::from("Hello");
    print_addr("String s1", &s1);
    println!("  s1 = {}", s1);
    
    println!("\nTransferring ownership:");
    let s2 = s1;
    print_addr("String s2 (same heap data)", &s2);
    println!("  s2 = {}", s2);
    println!("  s1 is no longer valid (moved)");
    // println!("{}", s1);  // COMPILE ERROR!
    
    println!("\nCloning to keep both:");
    let s3 = s2.clone();
    print_addr("String s3 (new heap data)", &s3);
    println!("  s2 = {}, s3 = {}", s2, s3);
    println!("  Both s2 and s3 are valid");
}

// Example 4: Borrowing - immutable and mutable
fn borrowing_example() {
    println!("\n=== 4. Borrowing - References ===");
    
    let mut data = vec![1, 2, 3, 4, 5];
    print_addr("Vector (stack)", &data);
    print_addr("Vector data (heap)", unsafe { &*data.as_ptr() });
    
    println!("\nImmutable borrow:");
    let r1 = &data;
    let r2 = &data;
    println!("  r1[0] = {}, r2[0] = {}", r1[0], r2[0]);
    println!("  Original still accessible: data[0] = {}", data[0]);
    
    println!("\nMutable borrow:");
    let r = &mut data;
    r[0] = 99;
    println!("  Modified via mutable reference: r[0] = {}", r[0]);
    // println!("{}", data);  // COMPILE ERROR while mutable borrow exists!
    
    println!("\nAfter mutable borrow ends:");
    println!("  data[0] = {}", data[0]);
}

// Example 5: Function call stack
fn function_c() {
    println!("  In function_c");
    let c = 3;
    print_addr("Variable c in function_c", &c);
    println!("  function_c returning...");
}

fn function_b() {
    println!("  In function_b");
    let b = 2;
    print_addr("Variable b in function_b", &b);
    function_c();
    println!("  Back in function_b, b still valid: {}", b);
    println!("  function_b returning...");
}

fn function_a() {
    println!("  In function_a");
    let a = 1;
    print_addr("Variable a in function_a", &a);
    function_b();
    println!("  Back in function_a, a still valid: {}", a);
    println!("  function_a returning...");
}

fn function_call_example() {
    println!("\n=== 5. Function Call Stack ===");
    println!("Calling nested functions (watch stack addresses):");
    function_a();
    println!("All functions returned, stack unwound");
}

// Example 6: Array/Vector allocation
fn array_allocation_example() {
    println!("\n=== 6. Array/Collection Allocation ===");
    
    println!("Stack array:");
    let stack_array = [1, 2, 3, 4, 5];
    print_addr("stack_array (stack)", &stack_array);
    println!("  Using: stack_array[0] = {}", stack_array[0]);
    let mut stack_array = stack_array;
    stack_array[0] = 99;
    println!("  After modification: stack_array[0] = {}", stack_array[0]);
    
    println!("\nVec (heap-allocated):");
    let mut vec = vec![1, 2, 3, 4, 5];
    print_addr("Vec object (stack)", &vec);
    print_addr("Vec data (heap)", unsafe { &*vec.as_ptr() });
    println!("  Using: vec[0] = {}", vec[0]);
    vec[0] = 99;
    println!("  After modification: vec[0] = {}", vec[0]);
    
    println!("\nBoxed array (heap):");
    let mut boxed = Box::new([1, 2, 3, 4, 5]);
    print_addr("Box (stack)", &boxed);
    print_addr("Array data (heap)", &boxed[0]);
    println!("  Using: boxed[0] = {}", boxed[0]);
    boxed[0] = 99;
    println!("  After modification: boxed[0] = {}", boxed[0]);
    
    println!("\nAll cleaned up automatically at scope end");
}

// Example 7: Lifetime tracking
fn lifetime_example() {
    println!("\n=== 7. Lifetime Tracking ===");
    
    let outer = String::from("Outer");
    print_addr("Outer string", &outer);
    println!("  outer = {}", outer);
    
    {
        let inner = String::from("Inner");
        print_addr("Inner string", &inner);
        println!("  In inner scope: outer = {}, inner = {}", outer, inner);
    }  // inner dropped here
    
    println!("  After inner scope: outer = {}", outer);
    println!("  inner is no longer accessible");
}

// Example 8: Reference counting
fn shared_ownership_example() {
    println!("\n=== 8. Shared Ownership with Rc ===");
    
    println!("Creating Rc:");
    let data = Rc::new(DataObject::new(10));
    print_addr("Rc (stack)", &data);
    println!("  Strong count: {}", Rc::strong_count(&data));
    
    {
        println!("\nCloning Rc (shares ownership):");
        let data2 = Rc::clone(&data);
        print_addr("Cloned Rc (stack)", &data2);
        println!("  Strong count: {}", Rc::strong_count(&data));
        println!("  Both point to same heap data");
        println!("  data id: {}, data2 id: {}", data.get_id(), data2.get_id());
    }  // data2 dropped, count decreases
    
    println!("\nAfter data2 scope:");
    println!("  Strong count: {}", Rc::strong_count(&data));
    println!("  data still valid: id = {}", data.get_id());
    
    println!("\nEnd of function, last Rc dropped...");
}

// Example 9: Copy vs Move semantics
fn copy_vs_move_example() {
    println!("\n=== 9. Copy vs Move Semantics ===");
    
    println!("Copy trait (stack primitives):");
    let x = 42;
    print_addr("x", &x);
    let y = x;  // Copied
    print_addr("y", &y);
    println!("  x = {}, y = {} (both valid)", x, y);
    
    println!("\nMove semantics (heap types):");
    let s1 = String::from("Hello");
    print_addr("s1", &s1);
    let s2 = s1;  // Moved
    print_addr("s2", &s2);
    println!("  s2 = {}", s2);
    println!("  s1 is no longer accessible (moved)");
    // println!("{}", s1);  // COMPILE ERROR!
}

// Example 10: Scope-based cleanup
fn scope_cleanup_example() {
    println!("\n=== 10. Scope-Based Resource Cleanup ===");
    
    println!("Outer scope:");
    let obj1 = DataObject::new(100);
    println!("  Created obj1");
    
    {
        println!("\n  Inner scope:");
        let _obj2 = DataObject::new(200);
        println!("    Created obj2");
        println!("    Both obj1 and obj2 valid");
        println!("  Inner scope ending...");
    }  // obj2 dropped here
    
    println!("\nBack in outer scope:");
    println!("  obj1 still valid: id = {}", obj1.get_id());
    println!("Outer scope ending...");
}  // obj1 dropped here

fn main() {
    println!("=== Comprehensive Memory Usage Patterns in Rust ===");
    println!("Demonstrating allocation -> use -> deallocation for each pattern\n");
    
    stack_allocation_example();
    heap_allocation_example();
    ownership_transfer_example();
    borrowing_example();
    function_call_example();
    array_allocation_example();
    lifetime_example();
    shared_ownership_example();
    copy_vs_move_example();
    scope_cleanup_example();
    
    println!("\n=== Summary ===");
    println!("Stack: Allocate -> Use -> Auto-cleanup (LIFO, Copy semantics)");
    println!("Heap:  Allocate -> Use -> Auto-cleanup via ownership/RAII");
    println!("       Rust enforces memory safety at compile time!");
    println!("       No use-after-free, no double-free, no data races");
}
